File: util/circular-chk.js
const fs = require("fs");
const strict = process.argv[2] === "--strict";
const walk = function(dir, filter) {
    var results = [];
    var list = fs.readdirSync(dir);
    list.forEach(function(file) {
        file = dir + "/" + file;
        var stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            /* Recurse into a subdirectory */
            results = results.concat(walk(file, filter));
        } else {
            /* Is a file */
            if (file.toLowerCase().endsWith(filter.toLowerCase())) {
                results.push(file);
            }
        }
    });
    return results;
};
const regexExclude = /keymaps\/(.*)via\//g;
const containsSRC = [];
const filesToScan = walk("./keyboards", "rules.mk");
for (const f of filesToScan) {
    if(regexExclude.test(f)){
        continue;
    }
    let content = fs.readFileSync(f, "utf-8");
    content = content.replace(/\\\n/g, "");
    const lines = content.split("\n");
    for (const l of lines) {
        if (l.startsWith("SRC =")) {
            const basepath = f.replace("./keyboards/", "");
            const imports = l
                .replace("SRC =", "")
                .replace(/\t/g, "")
                .split(" ");
            for (const i of imports) {
                const p = i.split("/");
                if (p[0] === "keyboards" && p[1] !== basepath.split("/")[0]) {
                    containsSRC.push(f);
                }
            }
        }
    }
}
if (containsSRC.length > 0) {
    console.log(JSON.stringify(containsSRC));
    console.log(`${containsSRC.length} files contains SRC illegal refs`);
    if (strict) {
        console.log(`Error!!`);
        process.exit(1);
    } else {
        console.log(`Non strict, exiting silently!`);
    }
} else {
    console.log(`All files are OK!`);
}

File: util/circular-chk.py
import glob
import re
from sys import exit, argv
from pathlib import Path

strict = False
if len(argv) > 1 and argv[1] == '--strict' :
    strict = True

regexExclude = r"keymaps\/(.*)via\/"
reg = re.compile(regexExclude)
filesToscan = Path('keyboards').glob('**/rules.mk')
illegalSRC = list()

for f in filesToscan:
    p = str(f).replace('\\','/')
    basepath = p.replace('keyboards/','')
    basepath_arr = basepath.split('/')

    # Skip 'via' file
    if reg.search(p):
        continue

    # Read file as lines
    lineList = list()
    with f.open() as file:
        try:
            lineList = file.readlines()
        except:
            # sometimes encoding issues with japanese characters for example
            pass

    # Concat multilines
    lineList = '\n'.join(lineList).replace('\\\n','').split('\n')
    for l in lineList:
        if not l.startswith('SRC ='):
            continue
        imports = l.replace('SRC =','').replace('\t','').split(' ')
        for i in imports:
            _i = i.split('/')
            if(_i[0] == 'keyboards' and _i[1] != basepath_arr[0]):
                illegalSRC.append(f)

if len(illegalSRC) > 0:
    for fp in illegalSRC:
        print(str(fp))
    print(f'{len(illegalSRC)} files with illegal imports')
    if strict:
        exit(1)
    else:
        exit(0)
